### PR TITLE
Refactor/sync

### PR DESCRIPTION
**Pull request by sijumoncy**
#629  
- Refactor Sync Functionality
- New UI Implementation
- Fixing Bugs

### CODE REVIEW

File: renderer/src/components/Sync/Ag/SyncToGiteaUtils.js [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+      filepath: result.path,
+      content: fileContent,
+      message: `updated ${filePath}`,
+      author: {
+        email: auth.user.email,
+        username: auth.user.username,
+      },
+      sha: result.sha,
+    // eslint-disable-next-line no-unused-vars
+    }).then((res) => {
```
Comment (samueljd) [inline renderer/src/components/Sync/Ag/SyncToGiteaUtils.js]: async/await mixed with promises.

File: renderer/src/components/Sync/Ag/SyncToGiteaUtils.js [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+      logger.debug('SyncTOGiteaUtils.js', `failed to upload file to Gitea metadata.json : ${err}`);
+    });
+  });
+};
+
+// sync profile updation
+export const createSyncProfile = async (auth) => {
+  const fs = window.require('fs');
+  const path = require('path');
+  await localForage.getItem('userProfile').then(async (user) => {
```
Comment (samueljd) [inline renderer/src/components/Sync/Ag/SyncToGiteaUtils.js:97]: async/await mixed with promises.

File: renderer/src/components/Sync/Ag/SyncToGitea.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+export async function uploadToGitea(projectDataAg, auth, setSyncProgress, notifyStatus, addNotification) {
+    logger.debug('ToGiteaUtils.js', 'in uploadTOGitea');
+    const projectData = projectDataAg.projectMeta;
+    const projectId = Object.keys(projectData.identification.primary.ag)[0];
+    const projectName = projectData.identification.name.en;
+    const ingredientsObj = projectData.ingredients;
+    const projectCreated = projectData.meta.dateCreated.split('T')[0];
+    const repoName = `ag-${projectData.languages[0].tag}-${projectData.type.flavorType.flavor.name}-${projectName.replace(/[\s+ -]/g, '_')}`;
+
+    await localForage.getItem('userProfile').then(async (user) => {
```
Comment (samueljd) [inline renderer/src/components/Sync/Ag/SyncToGitea.js]: await .then

File: renderer/src/components/Sync/Ag/SyncToGiteaUtils.js [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    };
+    // eslint-disable-next-line no-async-promise-executor
+    return new Promise(async (resolve) => {
+      await createRepository(
+        {
+          config: auth.config,
+          repo: settings?.name,
+          settings,
+        },
+      ).then((result) => {
```
Comment (samueljd) [inline renderer/src/components/Sync/Ag/SyncToGiteaUtils.js:25]: await .then

File: renderer/src/components/Sync/Ag/SyncToGiteaUtils.js [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+      repo: repoName,
+      branch,
+      filepath: filePath,
+      content: fileContent,
+      message: `commit ${filePath}`,
+      author: {
+        email: auth.user.email,
+        username: auth.user.username,
+      },
+    }).then(() => {
```
Comment (samueljd) [inline renderer/src/components/Sync/Ag/SyncToGiteaUtils.js]: await .then

File: renderer/src/components/Sync/Ag/SyncToGiteaUtils.js [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+export const updateFiletoServer = async (fileContent, filePath, branch, repoName, auth) => {
+  await readContent(
+    {
+      config: auth.config,
+      owner: auth.user.login,
+      repo: repoName.toLowerCase(),
+      ref: branch,
+      filepath: filePath,
+    },
+  ).then(async (result) => {
```
Comment (samueljd) [inline renderer/src/components/Sync/Ag/SyncToGiteaUtils.js]: await .then

File: renderer/src/components/Sync/Ag/SyncToGiteaUtils.js (hunk 1 above)
Comment (samueljd) [inline renderer/src/components/Sync/Ag/SyncToGiteaUtils.js]: await .then

File: renderer/src/components/Sync/Ag/SyncToGitea.js (hunk 3 above)
Comment (samueljd) [inline renderer/src/components/Sync/Ag/SyncToGitea.js]: await .then

File: renderer/src/components/Sync/Ag/SyncToGiteaUtils.js (hunk 4 above)
Comment (samueljd) [inline renderer/src/components/Sync/Ag/SyncToGiteaUtils.js:25]: await .then

File: renderer/src/components/Sync/Ag/SyncToGiteaUtils.js (hunk 5 above)
Comment (samueljd) [inline renderer/src/components/Sync/Ag/SyncToGiteaUtils.js]: await .then

File: renderer/src/components/Sync/Gitea/EditorSync/EditorSyncUtils.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+import { environment } from '../../../../../environment';
+import * as logger from '../../../../logger';
+import {
+ createFiletoServer, getOrPutLastSyncInAgSettings, handleCreateRepo, updateFiletoServer,
+} from '../../Ag/SyncToGiteaUtils';
+import packageInfo from '../../../../../../package.json';
+
+export async function getGiteaUsersList() {
+  let usersList = [];
+  await localForage.getItem('userProfile').then(async (user) => {
```
Comment (samueljd) [inline renderer/src/components/Sync/Gitea/EditorSync/EditorSyncUtils.js]: await .then

File: renderer/src/components/Sync/Gitea/ProjectMerge/MergeActions.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+      });
+      const requestOptions = {
+        method: 'POST',
+        headers: myHeaders,
+        body: payloadPr,
+        redirect: 'follow',
+      };
+      // check merge possible or not
+      let returnData = null;
+      await fetch(urlPr, requestOptions).then((resp) => resp.json().then((data) => ({ resposne: resp, body: data })))
```
Comment (samueljd) [inline renderer/src/components/Sync/Gitea/ProjectMerge/MergeActions.js]: await .then

File: renderer/src/components/Sync/Gitea/ProjectMerge/MergeActions.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+      .then((async (result) => {
```
Comment (samueljd) [inline renderer/src/components/Sync/Gitea/ProjectMerge/MergeActions.js]: await .then

File: renderer/src/components/Sync/Gitea/ProjectMerge/MergeActions.js [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+            logger.debug('MergeActions.js', 'PR success - continue Merge operations');
+            actions.setStepCount((prevStepCount) => prevStepCount + 1);
+            const mergePayload = JSON.stringify({
+              Do: 'merge',
+              delete_branch_after_merge: false,
+              });
+            requestOptions.body = mergePayload;
+            const urlMerge = `${environment.GITEA_API_ENDPOINT}/repos/${selectedGiteaProject?.repo?.owner?.username}/${selectedGiteaProject?.repo?.name}/pulls/${result.body.number}/merge`;
+            // perform merge action
+            await fetch(urlMerge, requestOptions).then((response) => response)
```
Comment (samueljd) [inline renderer/src/components/Sync/Gitea/ProjectMerge/MergeActions.js]: await .then

File: renderer/src/components/Sync/Gitea/ProjectMerge/MergeActions.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+            .then(async (mergeResult) => {
```
Comment (samueljd) [inline renderer/src/components/Sync/Gitea/ProjectMerge/MergeActions.js]: .then

File: renderer/src/components/Sync/Gitea/ProjectMerge/MergeActions.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+                returnData = { status: 'success', message: 'Merge Success' };
+              } if (mergeResult.status === 405) {
+                logger.debug('MergeActions.js', 'Can not merge - nothing to merge or error ', mergeResult.resposne.statusText);
+                throw new Error(mergeResult.resposne.statusText);
+              }
+            });
+          } else {
+            // merge is not possible - conflict , Display Conflict Message
+            let htmlPart = '<div></div>';
+            await fetch(result.body.html_url).then((response) => response.text())
```
Comment (samueljd) [inline renderer/src/components/Sync/Gitea/ProjectMerge/MergeActions.js]: await .then

File: renderer/src/components/Sync/Gitea/ProjectMerge/ProjectMergeUtils.js [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+  myHeaders.append('Content-Type', 'application/json');
+  const requestOptions = {
+    method: 'DELETE',
+    headers: myHeaders,
+    body: '',
+    redirect: 'follow',
+  };
+  const urlDeleteBranch = `${GITEA_BASE_API_URL}/repos/${selectedGiteaProject?.repo?.owner?.username}/${selectedGiteaProject?.repo?.name}/branches/${selectedGiteaProject.branch.name}-merge`;
+  await fetch(urlDeleteBranch, requestOptions)
+  .then((response) => response)
```
Comment (samueljd) [inline renderer/src/components/Sync/Gitea/ProjectMerge/ProjectMergeUtils.js:56]: await .then

File: renderer/src/components/Sync/Gitea/SyncFromGitea.js [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+import * as logger from '../../../logger';
+import { environment } from '../../../../environment';
+import { importServerProject } from './SyncFromGiteaUtils';
+
+export async function downloadFromGitea(repo, auth, setSyncProgress, notifyStatus, setSelectedGiteaProject, addNotification) {
+  logger.debug('SyncFromGitea.js', 'in SyncFromGiea : onClick offline sync');
+  try {
+    const currentUser = await localForage.getItem('userProfile');
+    await fetch(`${environment.GITEA_API_ENDPOINT}/repos/${repo.owner.username}/${repo.name}/branches`)
+    .then((response) => response.json())
```
Comment (samueljd) [inline renderer/src/components/Sync/Gitea/SyncFromGitea.js]: await .then

File: renderer/src/components/Sync/Gitea/SyncFromGitea.js [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+            {
+            config: auth.config,
+            owner: auth.user.login,
+            repo: repo.name,
+            ref: foundSyncBranch?.name,
+            filepath: 'metadata.json',
+            },
+        );
+        await fetch(readMetaData.download_url).then((resp) => resp.json())
+        .then(async (metaFile) => {
```
Comment (samueljd) [inline renderer/src/components/Sync/Gitea/SyncFromGitea.js]: await .then

File: renderer/src/components/Sync/Gitea/SyncFromGiteaUtils.js [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+    await readContent(
+        {
+          config: auth.config,
+          owner: auth.user.login,
+          repo: repo.name,
+          ref: userBranch?.name,
+          filepath: key,
+        },
+      // eslint-disable-next-line no-loop-func
+      ).then(async (result) => {
```
Comment (samueljd) [inline renderer/src/components/Sync/Gitea/SyncFromGiteaUtils.js]: await .then

File: renderer/src/components/Sync/Gitea/SyncFromGiteaUtils.js [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+        logger.debug('giteaUtils import.js', 'sending the data from Gitea with content');
+        // console.log('file from server : ', result.name);
+        if (result !== null) {
+          await fetch(result.download_url)
+              .then((resposne) => resposne.text())
```
Comment (samueljd) [inline renderer/src/components/Sync/Gitea/SyncFromGiteaUtils.js]: await .then

File: renderer/src/components/Sync/hooks/useGetCurrentProjectMeta.js [hunk 18]
```diff
@@ -0,0 +1,32 @@
+import fetchProjectsMeta from '@/core/projects/fetchProjectsMeta';
+import localforage from 'localforage';
+import { useState } from 'react';
+
+export default function useGetCurrentProjectMeta() {
+    let meta;
+    const [currentProjectMeta, setCurrentProjectMeta] = useState(undefined);
+
+    const getProjectMeta = async (currentProjectName) => {
+        await localforage.getItem('userProfile').then(async (user) => {
```
Comment (samueljd) [inline renderer/src/components/Sync/hooks/useGetCurrentProjectMeta.js]: await .then

File: renderer/src/components/Sync/useSync.js [hunk 19]
```diff
@@ -1,137 +1,81 @@
...
+    syncStarted: false,
+    totalFiles: 0,
+    completedFiles: 0,
+  });
+  const [refreshGiteaListUI, setRefreshGiteaListUI] = useState({ triger: false, timeOut: false });
 
   const fetchProjects = async () => {
-    logger.debug('Dropzone.js', 'calling fetchProjects event');
+    logger.debug('UseSync.js', 'calling fetchProjects event');
     localForage.getItem('userProfile').then(async (user) => {
```
Comment (samueljd) [inline renderer/src/components/Sync/useSync.js]: async/await

File: renderer/src/components/Sync/useSync.js [hunk 20]
```diff
@@ -1,137 +1,81 @@
...
       await fetchProjectsMeta({ currentUser: user?.username })
-      .then((value) => {
-      // console.log("value :  " ,value.projects);
-      setAgProjectsMeta(value.projects);
-      value.projects.forEach((project) => {
-        projectList.push(project.identification.name.en);
-        // console.log("project name :  " ,project.identification.name.en);
-      });
+      .then(async (value) => {
```
Comment (samueljd) [inline renderer/src/components/Sync/useSync.js:32]: await .then

File: renderer/src/components/hooks/useAddNotification.js [hunk 21]
```diff
@@ -10,8 +10,8 @@ export default function useAddNotification() {
         },
       } = useContext(AutographaContext);
 
-    const addNotification = async (title, text, type) => {
-        localforage.getItem('notification').then((value) => {
+    async function addNotification(title, text, type) {
+        await localforage.getItem('notification').then((value) => {
```
Comment (samueljd) [inline renderer/src/components/hooks/useAddNotification.js]: async/await

**Review by samueljd — CHANGES_REQUESTED** [review]
Please check the async await and promises mixups. 
In case of the need to use `.then` in an sync function remove the await since that is redundant.

**Review by vipinpaul — APPROVED** [review]

**Pull request merged**